lib/std/fs.zig

```diff
@@ -118,38 +118,30 @@ pub fn copyFileAbsolute(source_path: []const u8, dest_path: []const u8, args: Co
 /// TODO update this API to avoid a getrandom syscall for every operation.
 pub const AtomicFile = struct {
     file: File,
-    tmp_path_buf: [MAX_PATH_BYTES - 1:0]u8,
+    // TODO either replace this with rand_buf or use []u16 on Windows
+    tmp_path_buf: [TMP_PATH_LEN:0]u8,
     dest_path: []const u8,
     file_open: bool,
     file_exists: bool,
+    close_dir_on_deinit: bool,
     dir: Dir,
 
     const InitError = File.OpenError;
 
+    const TMP_PATH_LEN = base64.Base64Encoder.calcSize(12);
+
     /// TODO rename this. Callers should go through Dir API
-    pub fn init2(dest_path: []const u8, mode: File.Mode, dir: Dir) InitError!AtomicFile {
-        const dirname = path.dirname(dest_path);
+    pub fn init2(dest_path: []const u8, mode: File.Mode, dir: Dir, close_dir_on_deinit: bool) InitError!AtomicFile {
         var rand_buf: [12]u8 = undefined;
-        const dirname_component_len = if (dirname) |d| d.len + 1 else 0;
-        const encoded_rand_len = comptime base64.Base64Encoder.calcSize(rand_buf.len);
-        const tmp_path_len = dirname_component_len + encoded_rand_len;
-        var tmp_path_buf: [MAX_PATH_BYTES - 1:0]u8 = undefined;
-        if (tmp_path_len > tmp_path_buf.len) return error.NameTooLong;
-
-        if (dirname) |dn| {
-            mem.copy(u8, tmp_path_buf[0..], dn);
-            tmp_path_buf[dn.len] = path.sep;
-        }
-
-        tmp_path_buf[tmp_path_len] = 0;
-        const tmp_path_slice = tmp_path_buf[0..tmp_path_len :0];
+        var tmp_path_buf: [TMP_PATH_LEN:0]u8 = undefined;
+        tmp_path_buf[base64.Base64Encoder.calcSize(12)] = 0;
```

Comment:
This feels suboptimal to me and I'd prefer to use
```suggestion
        tmp_path_buf[TMP_PATH_LEN] = 0;
```
but I got "cannot assign to constant" errors when I tried that and variants. I don't know a way to make this cleaner (I'm still fairly new to Zig, sorry).

Reply:
hmm to be honest that sounds like a compiler bug. I'll take a look